Source/DafnyCompilers/CompilerRewriter.dfy

```diff
@@ -0,0 +1,2215 @@
+      : (r: PureInterpResult<WV>)
+      requires fn.Closure?
+      requires |fn.vars| == |argvs|
+    // Small utility, very similar to ``InterpFunctionCall``, which we use to factorize
+    // the definitions. The opaque attribute is maybe not necessary anymore, but as the
+    // proofs work with it, we might as well keep it (it is easier to remove an opaque
+    // attribute, than to add one and fix the proofs by adding the proper calls to ``reveal``).
+    {
+        var ctx := BuildCallState(fn.ctx, fn.vars, argvs);
+        var Return(val, ctx) :- InterpExpr(fn.body, fuel, ctx);
+        Success(val)
+    }
+
+    predicate {:verify true} EqValue(v: WV, v': WV)
+      decreases ValueTypeHeight(v) + ValueTypeHeight(v'), 1
+    // Equivalence between values.
+    // 
+    // Two values are equivalent if:
+    // - they are not closures and are equal/have equivalent children values
+    // - they are closures and, when applied to equivalent inputs, they return equivalent outputs
+    //
+    // Rk.: we could write the predicate in a simpler manner by using `==` in case the values are not
+    // closures, but we prepare the terrain for a more general handling of collections.
+    //
+    // Rk.: for now, we assume the termination. This function terminates because the size of the
+    // type of the values decreases, the interesting case being the closures (see ``EqValue_Closure``).
+    // Whenever we find a closure `fn_ty = (ty_0, ..., ty_n) -> ret_ty`, we need to call ``EqValue``
+    // on valid inputs (with types `ty_i < fn_ty`) and on its output (with type `ret_ty < fn_ty`).
+    //
+    // Rk.: I initially wanted to make the definition opaque to prevent context saturation, because
+    // in most situations we don't need to know the content of EqValue.
+    // However it made me run into the following issue:
+    // BUG(https://github.com/dafny-lang/dafny/issues/2260)
+    // As ``EqValue`` appears a lot in foralls, using the `reveal` trick seemed too cumbersome
+    // to be a valid option.
+    {
+      match (v, v') {
+        case (Bool(b), Bool(b')) => b == b'
+        case (Char(c), Char(c')) => c == c'
+        case (Int(i), Int(i')) => i == i'
+        case (Real(r), Real(r')) => r == r'
+        case (BigOrdinal(o), BigOrdinal(o')) => o == o'
+        case (BitVector(width, value), BitVector(width', value')) =>
+          && width == width'
+          && value == value'
+        case (Map(m), Map(m')) =>
+          ValueTypeHeight_Children_Lem(v); // For termination
+          ValueTypeHeight_Children_Lem(v'); // For termination
+          && m.Keys == m'.Keys
+          && |m| == |m'| // We *do* need this
+          && (forall x | x in m :: EqValue(m[x], m'[x]))
+        case (Multiset(ms), Multiset(ms')) =>
+          ms == ms'
+        case (Seq(sq), Seq(sq')) =>
+          ValueTypeHeight_Children_Lem(v); // For termination
```

Comment:
In general lemmas in Dafny are called `LemmaXYZ` (not a fan, but https://github.com/dafny-lang/libraries/blob/master/STYLE.md)